reporter: Report checkup results

orelmisan opened the pull request:
Until now, the results reported by the checkup hasn't been read by the framework.

Read and report the checkup's reported results:
1. `succeeded`.
2. `failureReason`.
3. Checkup custom results (optional).

The "succeeded" field value is a "true", when no errors had occurred in the framework or in the checkup.

The `failureReason` field is a concatenation of:
1. The value reported by the checkup (if exists).
2. `Checkup.Run()` error (if occurred).
3. `Checkup.Results()` error (if occurred).
4. `Checkup.Teardown()` error (if occurred).

~~Depends on PR #46.~~

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/9a2eb2fdf45b5b3777f15f2808b5f5bd091b1c6b..b326b10de502e48fbdecf7405c638b6b90a17749): Mistakenly changed a commit message from PR #46.

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/b326b10de502e48fbdecf7405c638b6b90a17749..7a10359fe235d032970a2cc553c22ebba60f4a66): Rebase.

EdDev commented on kiagnose/internal/results/results.go:
```diff
@@ -0,0 +1,92 @@
+	resultsData := Results{}
+
+	if resultsData.Succeeded, err = parseSucceededField(configMap.Data); err != nil {
+		return Results{}, err
```
nit: You could just return `resultsData` and avoid unneeded heap allocation.

orelmisan replied:
Done.

EdDev commented on kiagnose/internal/results/results.go:
```diff
@@ -0,0 +1,92 @@
+	}
+
+	if resultsData.FailureReason, err = parseFailureReasonField(configMap.Data); err != nil {
+		return Results{}, err
```
Same

orelmisan replied:
Done.

EdDev commented on kiagnose/internal/results/results_test.go:
```diff
@@ -0,0 +1,150 @@
+	configMapName      = "checkup1"
+)
+
+func TestReadResultsShouldSucceed(t *testing.T) {
```
nit: `When` can be appended to the name to avoid repetition below.

orelmisan replied:
Done.

EdDev commented on kiagnose/internal/results/results_test.go:
```diff
@@ -0,0 +1,150 @@
+	}
+}
+
+func TestReadResultsShouldFail(t *testing.T) {
```
nit: Same, `When` can be appended to the name.

orelmisan replied:
Done.

EdDev commented on kiagnose/internal/results/results_test.go:
```diff
@@ -0,0 +1,150 @@
+}
+
+func TestReadResultsShouldFail(t *testing.T) {
+	t.Run("when results fails to be fetched", func(t *testing.T) {
```
Isn't `failing to fetch results` simpler?

orelmisan replied:
Thanks, done.

EdDev reviewed (changes requested):
> Report "succeeded" and "failureReason" fields to the user-supplied
ConfigMap object.

Is the code in this commit dependent on the user ConfgiMap? I hope not.
If it is not, it should not mention something as specific as that, it is confusing.

> logical AND

Please use a less technical term. Try to tell a story that describes what you intended it to be.

You also have a numeration issue (point 3 appears twice).

EdDev reviewed (commented):
> checkupStub's methods now have a pointer receiver, as for the linter's
suggestion, due to its size.

:/

EdDev commented on kiagnose/internal/launcher/launcher_test.go:
```diff
@@ -45,26 +45,47 @@ const (
 
-func TestLauncherRunsSuccessfully(t *testing.T) {
-	fakeClient := fake.NewSimpleClientset(newConfigMap(checkupSpecData()))
+func TestLauncherRunWithResults(t *testing.T) {
```
I do not think we need to mention it runs with "results".
This change has two parts:
- Use `Results` in the existing test.
- Add a new test.

One can even claim there is a 3rd part: Convert it to a table the existing.

Splitting like this also helps reason for it and to review the change (current diff is not that nice).

EdDev replied:
Forget it.

EdDev reviewed (commented):
Second commit.

EdDev reviewed (commented):
> results: Read and report checkup partial results 

What does it mean "partial"?
I got confused with the last commit.

orelmisan commented:
> > Report "succeeded" and "failureReason" fields to the user-supplied
> > ConfigMap object.
> 
> Is the code in this commit dependent on the user ConfgiMap? I hope not. If it is not, it should not mention something as specific as that, it is confusing.

The code is not ConfigMap dependent.
When we'll transition to a CR, only the Reporter and the launcher tests will change.
The commit message was updated.

> > logical AND
> 
> Please use a less technical term. Try to tell a story that describes what you intended it to be.

Done.
 
> You also have a numeration issue (point 3 appears twice).

Thanks.

EdDev reviewed (commented):
> Reporter.Report() now receives a pointer to Status, as for the linter's
suggestion, due to its size.

Please extract this pointer stuff out of the rest of the change. It makes it harder to review.

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/c028e05341669aaca582c3529d45ba9b086e6e86..5776e7a3087accdc6b93a345edcba95c685ebd97): Updated first commit message.

EdDev commented on kiagnose/internal/launcher/launcher_test.go:
```diff
@@ -45,26 +45,47 @@ const (
-func TestLauncherRunsSuccessfully(t *testing.T) {
-	fakeClient := fake.NewSimpleClientset(newConfigMap(checkupSpecData()))
+func TestLauncherRunWithResults(t *testing.T) {
+	const testFailureReason = "some reason"
```
nit: I do not think this is need. Is the linter shouting without it?

orelmisan replied:
I think we can do without the const.

orelmisan replied:
Done.

orelmisan commented:
> > results: Read and report checkup partial results
> 
> What does it mean "partial"? I got confused with the last commit.

The first commit reads and reports the checkup's "succeeded" and "failureReason" fields.
The third commit also reads the checkup's custom results.

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/5776e7a3087accdc6b93a345edcba95c685ebd97..b321b1fcb12b9ee83f9c26178f9e39c36a38cafb): Answered review comments:
1. Added a commit for the change in Reporter.Report().
2. Changed commit messages.
3. Removed redundant const in launcher tests.

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/b321b1fcb12b9ee83f9c26178f9e39c36a38cafb..154dbb12ccc9416ef8ab54b0247f3d9013f7ca87): Removed redundant "when" from results test.

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/154dbb12ccc9416ef8ab54b0247f3d9013f7ca87..64e953dbcf0d3937f1dd0c50348ce66aa62fd89f): launcher tests: Moved "when" from case description to function name.

EdDev reviewed (commented):
> reporter: Change Reporter.Report()

The subject line is suppose to tell me what was done here at high level.
The current one tells me that some change was done, but I have no clue what and that is not informative.

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/64e953dbcf0d3937f1dd0c50348ce66aa62fd89f..f7448376194711891c7c33397279bc9be0e99362): Changed commit message.

orelmisan commented:
> > reporter: Change Reporter.Report()
> 
> The subject line is suppose to tell me what was done here at high level. The current one tells me that some change was done, but I have no clue what and that is not informative.

Changed the title to `reporter: Change Reporter.Report() input type`.

ormergi reviewed (commented):
Hi, thanks for the PR looks pretty good.
Please see my inline comments most of them are suggestion to make the code easier to understand.

Regarding "reporter: Change Reporter.Report() input type"
I think those changes should be introduced by the commit that changes `Status` and cause the linert to fail.
BTW`Status` should be considered as reference type and thus addressed as pointer, not due to its size (same goes for the first commit message).

orelmisan commented on kiagnose/internal/launcher/launcher_test.go:
```diff
@@ -192,18 +200,28 @@ var (
 	failRun      error
+	failResults  error
 	failTeardown error
+	results      results.Results
```
Can you please elaborate?

ormergi replied:
`Resuts` holds time.Time and a map, both are reference types, if it passed to func as is it will be shallow copied.

EdDev replied:
I am not familiar by such a rule or best practice.
If there is no reason to have a pointer, I prefer not to have it as a default.

BTW, passing a struct by value gives no guaranties to the reader about its content not being mutated, for the reason you just mentioned (it can include directly or indirectly referenced types). But it does provides **intent**, which is no less valuable.

orelmisan commented on kiagnose/internal/launcher/launcher_test.go:
```diff
@@ -45,26 +45,45 @@ const (
+func TestLauncherRunWithResultsWhen(t *testing.T) {
+	type resultsTestCase struct {
+		description  string
+		inputResults results.Results
```
These are the results allegedly reported by the checkup.
They are compared against the `expectedResults` calculated bellow.

ormergi replied:
I understand what it does, I suggest to rename it so it will be easier to understand.

orelmisan replied:
Do you have another naming suggestion?

orelmisan commented on kiagnose/internal/launcher/launcher_test.go:
```diff
@@ -45,26 +45,45 @@ const (
+		},
+		{
+			description:  "checkup fails",
+			inputResults: results.Results{FailureReason: "some reason"},
```
In the success scenario, the `FailureReason` field is an empty string.
In the failure scenario, the `Succeeded` field is false.

There are the zero values of their types.
It look redundant to me.

orelmisan commented on kiagnose/internal/launcher/launcher.go:
```diff
@@ -78,6 +83,17 @@ func (l Launcher) Run() (runErr error) {
+			statusData.Succeeded = checkupResults.Succeeded
+			statusData.FailureReason = checkupResults.FailureReason
+		}
+
```
Lets continue this discussion [here](https://github.com/kiagnose/kiagnose/pull/48#discussion_r869208618).

orelmisan commented:
[Change](https://github.com/kiagnose/kiagnose/compare/248d5512ef62f918914895f7311914b0f0ab745b..e5503fa2f8a471961529b72dab561d4033178165): Fixed a small nit in launcher tests.

orelmisan commented on kiagnose/internal/launcher/launcher.go:
```diff
@@ -56,10 +58,13 @@ func (l Launcher) Run() (runErr error) {
-			statusData.Succeeded = true
-		} else {
-			statusData.FailureReason = runErr.Error()
+		if runErr != nil {
+			statusData.Succeeded = false
+			if statusData.FailureReason != "" {
+				statusData.FailureReason = fmt.Sprintf("%s, %v", statusData.FailureReason, runErr)
+			} else {
+				statusData.FailureReason = runErr.Error()
+			}
```
> Regarding "reporter: Change Reporter.Report() input type" I think those changes should be introduced by the commit that changes `Status` and cause the linter to fail.

@ormergi there was a previous [request](https://github.com/kiagnose/kiagnose/pull/48#pullrequestreview-967640028) to separate it, in order to ease the review.
It added a lot of noise and distracted from the main change.

orelmisan replied:
My reasoning for postponing it to a followup PR is the time it takes to manually perform the E2E tests and to make sure the change works + the time discussing the new solution.

This change works, it was tested against a cluster, its not optimal and it can be changed.

EdDev replied:
As originally stated in this thread, it can be handled in a following PR to allow progress.
I will try to handle it after this gets merged.

EdDev reviewed (approved):
I think this PR is in a good shape and we can take it as is and improve its code in following changes.

I do not like that we have been forced to change some usages to a pointer instead of a plain structure due to the linter.
I suggested #51 so we can revert these changes.
